heartbeat/monitors/job.go

```diff
@@ -21,10 +21,89 @@ import (
 	"github.com/elastic/beats/libbeat/beat"
 )
 
-// Job represents the work done by a single check by a given Monitor.
+// A Job represents a unit of execution, and may return multiple continuation jobs.
 type Job interface {
 	Name() string
-	Run() (beat.Event, []jobRunner, error)
+	Run() (*beat.Event, []Job, error)
 }
 
-type jobRunner func() (beat.Event, []jobRunner, error)
+// NamedJob represents a job with an explicitly specified name.
+type NamedJob struct {
+	name string
+	run  func() (*beat.Event, []Job, error)
+}
+
+// CreateNamedJob makes a new NamedJob.
+func CreateNamedJob(name string, run func() (*beat.Event, []Job, error)) *NamedJob {
+	return &NamedJob{name, run}
+}
+
+// Name returns the name of this job.
+func (f *NamedJob) Name() string {
+	return f.name
+}
+
+// Run executes the job.
+func (f *NamedJob) Run() (*beat.Event, []Job, error) {
+	return f.run()
+}
+
+// AnonJob represents a job with no assigned name, backed by just a function.
+type AnonJob func() (*beat.Event, []Job, error)
+
+// Name returns "" for AnonJob values.
+func (aj AnonJob) Name() string {
+	return ""
+}
+
+// Run executes the function.
+func (aj AnonJob) Run() (*beat.Event, []Job, error) {
+	return aj()
+}
+
+// AfterJob creates a wrapped version of the given Job that runs additional
+// code after the original Job, possibly altering return values.
+func AfterJob(j Job, after func(*beat.Event, []Job, error) (*beat.Event, []Job, error)) Job {
+
+	return CreateNamedJob(
+		j.Name(),
```

Comment:
The other option is to have name be a function rather than a field. Every time the name was asked for we'd invoke the child job. That would allow names to be more dynamic and lazy.

Reply:
OTOH, I don't know if we actually need that. This is def simpler.